typesafeconfig/typesafe_config.go

```diff
@@ -0,0 +1,340 @@
+		MatchName: func(mapKey, fieldName string) bool {
+			normalizedMapKey := strings.ToLower(mapKey)
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "-", "")
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "_", "")
+			return strings.ToLower(fieldName) == normalizedMapKey
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = decoder.Decode(untypedConfig)
+	return typeSafeConfig, err
+}
+
+func loadEnvironmentSources() map[string]interface{} {
+	config := make(map[string]interface{})
+	env := os.Environ()
+	for _, envVar := range env {
+		kvPair := strings.SplitN(envVar, "=", 2)
+		rawKey := kvPair[0]
+		value := kvPair[1]
+		key := strings.Split(rawKey, "_")
+		setValue(config, key, value)
+	}
+	return config
+}
+
+func setValue(config map[string]interface{}, key []string, value any) {
+	if len(key) == 1 {
+		config[key[0]] = value
+		return
+	}
+	cur, remaining := pop(key)
+	var nested map[string]interface{}
+	if config[cur] == nil {
+		nested = make(map[string]interface{})
+	} else {
+		curNested := config[cur]
+		unboxed, ok := curNested.(map[string]interface{})
+		if !ok {
+			nested = make(map[string]interface{})
+		} else {
+			nested = unboxed
+		}
+	}
+	config[cur] = nested
+	setValue(nested, remaining, value)
+}
+
+func pop[T any](array []T) (T, []T) {
+	return array[0], array[1:]
+}
+
+// mergeSources recursively left merges config sources, omitting any non-map values that are not one of: lists, numbers, or booleans
+// un-flattens keys before merging into new map
+func mergeSources(sources ...map[string]interface{}) map[string]interface{} {
+	m := make(map[string]interface{})
+	for _, unNormalizedSource := range sources {
+		source := normalizeKeys(unNormalizedSource)
+		// iterate through key and if the value is a map recurse, else set the key to the value if type is a number, list or boolean
+		for _, key := range maps.Keys(source) {
+			val := source[key]
+			cur := m[key]
+			if cur == nil {
+				m[key] = val
+				continue
+			}
+
+			curT := reflect.TypeOf(cur)
+			valT := reflect.TypeOf(val)
+			switch curT.Kind() {
+			case reflect.Map:
+				typedCur := cur.(map[string]interface{})
+				if valT.Kind() == reflect.Map {
+					typedVal := val.(map[string]interface{})
+					m[key] = mergeSources(typedCur, typedVal)
+				} else {
+					m[key] = val
+				}
+			case reflect.Array, reflect.String, reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Float32, reflect.Float64:
+				m[key] = val
+			}
+		}
+	}
+	return m
+}
+
+func resolveSecrets(config map[string]interface{}) error {
+	for _, key := range maps.Keys(config) {
+		val := config[key]
+		valT := reflect.TypeOf(val)
+		if valT.Kind() == reflect.Map {
+			if err := resolveSecrets(val.(map[string]interface{})); err != nil {
+				return err
+			}
+		}
+		if valT.Kind() == reflect.String && secrets.IsEncryptedSecret(val.(string)) {
+			d, err := secrets.NewDecrypter(context.TODO(), val.(string))
```

Comment:
I have a dislike of `context.TODO()` and think people should just accept that they're not going to go back and thread context through everything. I would just use `context.Background()`

Reply:
kk, I copy pasted this from the config package, but that isn't an excuse, I'll clean this up.